File: book2-algorithms-sedgewick/ch1/ex-1.3.10-infix-to-postfix.py
#
# [Sedgewick][Ch1][Python] Exercise 1.3.10 Infix to Postfix conversion
#
# $ python ex-1.3.10-infix-to-postfix.py
# .
# ----------------------------------------------------------------------
# Ran 1 test in 0.000s
#
# OK
#
# Example Trace:
# INPUT: ((A+B)*C)+(A+B)            EXPECTED: AB+C*AB++
# IN              STACK             POSTFIX_EXPR
# (               (
# (               (,(
# A                                 A
# +               (,(,+
# B                                 A,B
# )               (,                A,B,+
# *               (,*
# C                                 A,B,+,C
# )                                 A,B,+,C,*
# +               +
# (               +,(
# A                                 A,B,+,C,*,A
# +               +,(,+
# B                                 A,B,+,C,*,A,B
# )                                 A,B,+,C,*,A,B,+,+
import unittest


def is_operator(ch: str) -> bool:
    if ch == "+" or ch == "-" or ch == "*" or ch == "/":
        return True
    return False


def infix_to_postfix(infix_expr: str) -> str:
    precedences = {"+": 1, "-": 1, "*": 2, "/": 2}
    # holds operators and parentheses
    stack = []
    # holds the postfix expression
    postfix_expr = []
    for ch in infix_expr:
        if ch.isspace():
            continue
        if ch.isalnum():
            postfix_expr.append(ch)
        elif ch == "(":
            stack.append(ch)
        elif ch == ")":
            while len(stack) > 0 and stack[-1] != "(":
                postfix_expr.append(stack.pop())
            stack.pop()
        elif is_operator(ch):
            while (
                len(stack) > 0
                and stack[-1] != "("
                and precedences[stack[-1]] >= precedences[ch]
            ):
                postfix_expr.append(stack.pop())
            stack.append(ch)

    while len(stack) > 0:
        postfix_expr.append(stack.pop())

    return "".join(postfix_expr)


class InfixToPostfix(unittest.TestCase):
    def test_infix_to_postfix(self):
        test_cases = [
            ("((A+B)*C)", "AB+C*"),
            ("((A+B)*C)+(A+B)", "AB+C*AB++"),
            ("(A+(B*C))", "ABC*+"),
            ("A+B*C", "ABC*+"),
            ("A*B+C", "AB*C+"),
        ]
        for test_case in test_cases:
            infix_expr = test_case[0]
            expected_postfix_expr = test_case[1]
            actual_postfix_expr = infix_to_postfix(infix_expr)
            self.assertEqual(expected_postfix_expr, actual_postfix_expr)


if __name__ == "__main__":
    unittest.main()

File: book2-algorithms-sedgewick/ch1/s01_exercises_rust/src/ex_1_3_10_infix_to_postfix.rs
/*
[Sedgewick][Ch1][Rust] Exercise 1.3.10 Infix to Postfix conversion

Example Trace:
INPUT: ((A+B)*C)+(A+B)            EXPECTED: AB+C*AB++
IN              STACK             POSTFIX_EXPR
(               (
(               (,(
A                                 A
+               (,(,+
B                                 A,B
)               (,                A,B,+
*               (,*
C                                 A,B,+,C
)                                 A,B,+,C,*
+               +
(               +,(
A                                 A,B,+,C,*,A
+               +,(,+
B                                 A,B,+,C,*,A,B
)                                 A,B,+,C,*,A,B,+,+
*/

#[allow(dead_code)]
use std::collections::HashMap;

fn is_operator(ch: char) -> bool {
    ch == '+' || ch == '-' || ch == '*' || ch == '/'
}

fn infix_to_postfix(infix_expr: &str) -> String {
    let mut postfix_expr: Vec<char> = Vec::new();
    let mut stack: Vec<char> = Vec::new();

    let mut precedences = HashMap::new();
    precedences.insert('+', 1);
    precedences.insert('-', 1);
    precedences.insert('*', 2);
    precedences.insert('/', 2);

    for ch in infix_expr.chars() {
        if ch.is_whitespace() {
            continue;
        }
        if ch.is_alphanumeric() {
            postfix_expr.push(ch);
        } else if ch == '(' {
            stack.push(ch)
        } else if ch == ')' {
            while stack.len() > 0 && stack.last() != Some(&'(') {
                postfix_expr.push(stack.pop().unwrap());
            }
            stack.pop();
        } else if is_operator(ch) {
            while stack.len() > 0
                && stack.last() != Some(&'(')
                && precedences[stack.last().unwrap()] >= precedences[&ch]
            {
                postfix_expr.push(stack.pop().unwrap());
            }
            stack.push(ch);
        }
    }

    while stack.len() > 0 {
        postfix_expr.push(stack.pop().unwrap());
    }

    String::from_iter(postfix_expr.iter())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_infix_to_postfix() {
        let test_cases = [
            ("((A+B)*C)", "AB+C*"),
            ("((A+B)*C)+(A+B)", "AB+C*AB++"),
            ("(A+(B*C))", "ABC*+"),
            ("A+B*C", "ABC*+"),
            ("A*B+C", "AB*C+"),
        ];
        for (infix_expr, expected) in test_cases {
            let actual = infix_to_postfix(infix_expr);
            assert_eq!(actual, expected);
        }
    }
}
